src/marker.rs

```diff
@@ -25,153 +364,302 @@ pub fn marker_2_rectancle(
+            namespace.remove(&marker_id);
+        });
+    }
+
+    fn clear(&mut self) {
+        self._markers.clear();
+    }
+
+    fn clear_namespace(&mut self, marker_ns: String) -> Vec<i32> {
+        let mut res = Vec::new();
+        self._markers.entry(marker_ns).and_modify(|namespace| {
+            for marker in namespace.values() {
+                res.push(marker.id);
+            }
+            namespace.clear();
+        });
+        res
+    }
+
+    fn get_lines(&self) -> Vec<Line> {
+        let mut res = Vec::<Line>::new();
+        for namespace in self._markers.values() {
+            for marker in namespace.values() {
+                res.extend(marker.lines.to_vec());
+            }
         }
+        res
     }
+}
 
-    pub fn get_lines(&self) -> Vec<Line> {
-        self.lines.read().unwrap().to_vec()
+///Class that handles the lifecycle of the markers.
+///
+/// Markers that provide a lifecycle (i.e. not 0) need to be deleted from the container
+/// once the lifecycle is reached. To do this we add another container which has a timer.
+/// The timer collects jobs after a delay to delete markers from the markers container.
+///
+/// Because the timer uses guards, we also need to store them which is done with a dict
+/// and a list. A repeated task is given to the timer to consume the list of deleted
+/// marker and free the guards.
+struct MarkersLifecycle {
+    _markers_container: Arc<RwLock<TermvizMarkerContainer>>,
+    _deleted_markers: Arc<Mutex<Vec<(String, i32)>>>,
+    _guards: Arc<Mutex<BTreeMap<(String, i32), timer::Guard>>>,
+    _timer: Arc<Mutex<timer::Timer>>,
+    _cleaner_guard: timer::Guard,
+}
+
+impl MarkersLifecycle {
+    pub fn new(marker_container: TermvizMarkerContainer) -> MarkersLifecycle {
+        let timer = Arc::new(Mutex::new(timer::Timer::new()));
+        let deleted_markers = Arc::new(Mutex::new(Vec::<(String, i32)>::new()));
+        let guards = Arc::new(Mutex::new(BTreeMap::<(String, i32), timer::Guard>::new()));
+
+        let clean_job_delete_markers_ref = deleted_markers.clone();
+        let clean_job_guards_ref = guards.clone();
+
+        let cleaner_guard =
+            timer
+                .lock()
+                .unwrap()
+                .schedule_repeating(chrono::Duration::seconds(5), move || {
+                    // every 5s clear the guards that outlived their markers
+                    let mut clean_job_delete_markers = clean_job_delete_markers_ref.lock().unwrap();
+                    let mut clean_job_guards = clean_job_guards_ref.lock().unwrap();
+                    for key in clean_job_delete_markers.iter() {
+                        clean_job_guards.remove(key);
+                    }
+                    clean_job_delete_markers.clear();
+                });
+
+        Self {
+            _markers_container: Arc::new(RwLock::new(marker_container)),
+            _deleted_markers: deleted_markers,
+            _guards: guards,
+            _timer: timer,
```

Comment:
do we need to store the timer, would it not be enough to store the guards (they should keep the timer alive)

Reply:
I still need a reference to the timer to schedule new tasks